packages/shadycss/src/shadow-parts.js

```diff
@@ -359,9 +383,239 @@ export function onInsertBefore(parentNode, newNode, referenceNode) {
+ * will be dead-code removed.
+ * @record
+ */
+class PartRulesMapEntry {
+  constructor() {
+    /**
+     * The ShadyCSS StyleNode for this ::part rule.
+     * @type {!StyleNode}
+     */
+    this.styleNode;
+
+    /**
+     * The CSS custom property names consumed by this ::part rule.
+     * @type {!Array<!string>}
+     */
+    this.consumedProperties;
+
+    /**
+     * The parsed part names that this ::part rule targets.
+     * @type {!Array<!string>}
+     */
+    this.partNames;
+  }
+}
+/* eslint-enable no-unused-vars */
+
+/**
+ * Look for any ::part rules in the given AST that consume custom properties,
+ * and load those into the part rules map.
+ *
+ * @param {!string} providerScope Lower-case tag name of the element whose
+ *     template this is.
+ * @param {!StyleNode} styleAst Parsed CSS for this template.
+ */
+export function analyzeTemplatePartRules(providerScope, styleAst) {
+  if (nativeCssVariables || !styleAst.rules) {
+    return;
+  }
+  for (const styleNode of styleAst.rules) {
+    const selector = styleNode['selector'];
+    if (!selector || selector.indexOf('::part') === -1) {
+      // TODO(aomarks) We do the `indexOf` check here to make the case where no
+      // `::part` rules are used is as fast as possible, but we could get away
+      // with just the `parsePartSelector` call we're doing next if the
+      // difference is negligible. Needs benchmarking.
+      continue;
+    }
+    const parsed = parsePartSelector(selector);
+    if (parsed === null) {
+      continue;
+    }
+    const consumedProperties = findConsumedCustomProperties(
+      styleNode['cssText']
+    );
+    if (consumedProperties.length === 0) {
+      continue;
+    }
+    const {elementName: receiverScope, parts} = parsed;
+    if (parts.length === 0) {
+      continue;
+    }
+    const key = [providerScope, receiverScope].join(':');
+    let entries = partRulesMap.get(key);
+    if (entries === undefined) {
+      entries = [];
+      partRulesMap.set(key, entries);
+    }
+    entries.push({
+      styleNode,
+      consumedProperties,
+      partNames: splitPartString(parts),
+    });
+  }
+}
+
+/**
+ * Extract the CSS custom property names that are consumed by the given CSS
+ * text.
+ *
+ * @param {!string} cssText The CSS text.
+ * @return {!Array<!string>}>} The CSS Custom Property names.
+ */
+function findConsumedCustomProperties(cssText) {
+  const obj = {};
+  StyleProperties.collectPropertiesInCssText(cssText, obj);
+  return Object.getOwnPropertyNames(obj);
+}
+
+/**
+ * Return entries from the partRulesMap that match the given scopes and all of
+ * the given part names.
+ */
+function lookupMatchingPartRules(scopeKey, partNames) {
+  const entries = partRulesMap.get(scopeKey);
+  if (entries === undefined) {
+    return [];
+  }
+  const matching = [];
+  for (const entry of entries) {
+    let matches = true;
+    for (const partName of entry.partNames) {
+      if (partNames.indexOf(partName) === -1) {
```

Comment:
Not following this. Why is it ok to bail on a match if one of entry partNames is not in the partNames argument?

Reply:
Sorry, wasn't too clearly written. I renamed the function, and added some documentation.

What's happening here is that, we have some part node with some list of part names from the `part` attribute, and we're trying to figure out whether there are any `::part` rules that match it, so that we can do the special handling for custom properties. By the spec, the specifier `::part(foo bar)` only matches part names with both `foo` AND `bar`. So the bail here is because we need to check if the given part node has ALL of the parts specified in the `::part` specifier, and we bail as soon as one is missing.